src/block_cache/block_chunkcache.c

```diff
@@ -0,0 +1,561 @@
+     * chunks.
+     */
+
+    if ((chunk_size = __chunkcache_admit_size(session)) == 0)
+        return (ENOSPC);
+    WT_RET(__wt_calloc(session, 1, sizeof(WT_CHUNKCACHE_CHUNK), newchunk));
+
+    /* Convert the block offset to the offset of the enclosing chunk. */
+    (*newchunk)->chunk_offset = WT_CHUNK_OFFSET(chunkcache, offset);
+    /* Chunk cannot be larger than the file. */
+    (*newchunk)->chunk_size = WT_MIN(chunk_size, (size_t)(block->size - (*newchunk)->chunk_offset));
+
+    /* Part of the hash ID was populated by the caller, but we must set the offset. */
+    (*newchunk)->hash_id = *hash_id;
+    (*newchunk)->hash_id.offset = (*newchunk)->chunk_offset;
+    hash = __wt_hash_city64((void *)hash_id, sizeof(WT_CHUNKCACHE_HASHID));
+    (*newchunk)->bucket_id = (unsigned int)(hash % chunkcache->hashtable_size);
+
+    if ((ret = __chunkcache_alloc(session, *newchunk)) != 0) {
+        __wt_free(session, *newchunk);
+        return (ret);
+    }
+    __wt_verbose(session, WT_VERB_CHUNKCACHE, "allocate: %s(%u), offset=%" PRIu64 ", size=%" PRIu64,
+      (char *)&(*newchunk)->hash_id.objectname, (*newchunk)->hash_id.objectid,
+      (uint64_t)(*newchunk)->chunk_offset, (uint64_t)(*newchunk)->chunk_size);
+
+    return (0);
+}
+
+/*
+ * __chunkcache_free_chunk --
+ *     Free the memory occupied by the chunk and the metadata.
+ */
+static void
+__chunkcache_free_chunk(WT_SESSION_IMPL *session, WT_CHUNKCACHE_CHUNK *chunk)
+{
+    WT_CHUNKCACHE *chunkcache;
+
+    chunkcache = &S2C(session)->chunkcache;
+
+    (void)__wt_atomic_sub64(&chunkcache->bytes_used, chunk->chunk_size);
+    WT_STAT_CONN_DECRV(session, chunk_cache_bytes, chunk->chunk_size);
+    WT_STAT_CONN_DECR(session, chunk_cache_chunks);
+
+    if (chunkcache->type == WT_CHUNKCACHE_IN_VOLATILE_MEMORY)
+        __wt_free(session, chunk->chunk_memory);
+    else {
+#ifdef ENABLE_MEMKIND
+        memkind_free(chunkcache->memkind, chunk->chunk_memory);
+#else
+        __wt_err(session, EINVAL,
+          "Chunk cache requires libmemkind, unless it is configured to be in DRAM");
+#endif
+    }
+    __wt_free(session, chunk);
+}
+
+/*
+ * __chunkcache_make_hash --
+ *     Populate the hash data structure, which uniquely identifies the chunk, and return the hash
+ *     table bucket number corresponding to this hash.
+ */
+static inline unsigned int
+__chunkcache_make_hash(WT_CHUNKCACHE *chunkcache, WT_CHUNKCACHE_HASHID *hash_id, WT_BLOCK *block,
+  uint32_t objectid, wt_off_t offset)
+{
+    uint64_t hash;
+
+    WT_CLEAR(*hash_id);
+    hash_id->objectid = objectid;
+    memcpy(&hash_id->objectname, block->name, WT_MIN(strlen(block->name), WT_CHUNKCACHE_NAMEMAX));
+    hash_id->offset = WT_CHUNK_OFFSET(chunkcache, offset);
+    hash = __wt_hash_city64((void *)hash_id, sizeof(WT_CHUNKCACHE_HASHID));
+
+    /* Return the bucket ID. */
+    return (unsigned int)(hash % chunkcache->hashtable_size);
```

Comment:
Why not just have the function return a `uint64_t` and use that in all the cases instead of casting to `unsigned int`?